File: Sliding_Puzzle.cpp
//	Alejandro Elizondo

#include <iostream>
#include <time.h>
#include <string>

using namespace std;
// Displays the puzzle
void Show (int Array[4][4]){
	cout<<"•---•---•---•---•---•"<<endl;
	for (int x = 0;x < 4;x++){
		cout << "| ";
		for (int y = 0;y < 4;y++){
			if (Array[x][y] < 10)
				cout << " " << Array[x][y] << " | ";
			else 
				cout << Array[x][y] << " | ";
		}
		cout << endl << "•---•---•---•---•---•" << endl;
	}	
}
// Generates a random puzzle
void Generate(int Array[4][4]){
	int iRand;
	string str = "";	// Used to keep count of used numbers

	for (int x = 0;x < 4;x++)
		for (int y = 0;y < 4;y++)
		{
				do{
					iRand = rand()%16;
				} while (str.find(iRand) != -1);
				Array[x][y] = iRand;
				str += iRand;
		}
}

int main() {
	int Array[4][4], iNum;
	srand(time(NULL));

	cout << "\t" << "\t" << "Welcome to the game"<< endl << endl;
    cout << "In order to win, order the numbers from 1 to 15, starting from the top-left corner. ";
    cout << "The bottom-right corner of the puzzle should be 0 in order to win."<< endl << endl;
	Generate(Array);
	Show(Array);
	do{
		cout << endl << "What number would you like to switch places with 0?" << endl;
		cin >> iNum;
		cout << iNum;
	} while (iNum < 1 || iNum > 15);
	return 0;
}

File: Java/Game/Sliding_Puzzle.java
//	Alejandro Elizondo
//	alexelmtz@outlook.com
//	Linkedin: linkedin.com/in/alejandro-elizondo
//	Github: github.com/alexelmtz/Sliding-Puzzle

package Game;

import java.util.*;

public class Sliding_Puzzle {

	public static void main(String[] args) {
		Scanner input = new Scanner(System.in);
		int iNum;
		boolean bPlay;
		
		System.out.println("\n Welcome to the game\n");
		System.out.print("In order to win, order the numbers from 1 to 15, starting \nfrom the top-left corner.");
		System.out.println("The bottom-right corner of the \npuzzle should be empty in order to win.\n");
		System.out.println("What game do you want to play? (type 3 or 4)");
		iNum = input.nextInt();
		GameBoard game = new GameBoard(iNum);

		game.Display();

		do {
			System.out.println("What number would you like to move?");
			iNum = input.nextInt();
			if (!game.Switch(iNum))
				System.out.println("Error that number can't be switched, please try again");
			game.Display();
			System.out.println("Do you want to keep playing? (true or false)");
			bPlay = input.nextBoolean();
		} while (bPlay);
		
		System.out.println("Thank you for playing");
		game.Display();
		input.close();
	}
}